File: ds4-gateway/media_resources_test.py
import importlib.util
from pathlib import Path
import unittest
from unittest.mock import patch
from collections import namedtuple

spec = importlib.util.spec_from_file_location('resources', Path(__file__).with_name('media_resources.py'))
resources = importlib.util.module_from_spec(spec)
spec.loader.exec_module(resources)


class Resources(unittest.TestCase):
    def test_host_facts_no_model_execution_and_no_private_paths(self):
        commands = []
        def command(*args):
            commands.append(args)
            if args[0] == 'nvidia-smi':
                return 'NVIDIA GB10'
            if args[:2] == ('docker', 'ps'):
                return ''
            if args[:2] == ('docker', 'info'):
                return '{"Architecture":"aarch64","DockerRootDir":"/private/docker"}'
            if args[:2] == ('docker', 'inspect'):
                return '[{"Mounts":[{"Type":"bind","Source":"/private/models","Destination":"/models"}]}]'
            self.fail('Unexpected command: ' + str(args))
        usage = namedtuple('usage', 'total used free')(1000, 400, 600)
        with patch.object(resources.platform, 'system', return_value='Linux'), patch.object(resources.platform, 'machine', return_value='aarch64'), patch.object(resources, 'command', side_effect=command), patch.object(resources.Path, 'read_text', return_value='MemTotal: 128000 kB\nMemAvailable: 4000 kB\n'), patch.object(resources.shutil, 'disk_usage', return_value=usage), patch.object(resources.os, 'stat', return_value=namedtuple('stat', 'st_dev')(42)):
            result = resources.collect({'container': 'a' * 64})
        self.assertEqual(result['memory_total_bytes'], 128000 * 1024)
        self.assertEqual(result['memory_available_bytes'], 4000 * 1024)
        self.assertEqual(len(result['disks']), 1)
        self.assertEqual(result['disks'][0]['locations'], ['home', 'docker', 'existing container mount: /models'])
        self.assertNotIn('/private/', str(result))
        self.assertEqual(len(commands), 4)
        self.assertTrue(all('exec' not in c and 'start' not in c and 'stop' not in c for c in commands))

    def test_inventory_filters_private_config_and_bounds_observation(self):
        import json
        ids = [format(i, '064x') for i in range(130)]
        calls = []
        def command(*args):
            calls.append(args)
            if args[1] == 'ps': return '\n'.join(ids)
            return json.dumps([{'Id': value, 'Image': 'sha256:' + 'a'*64,
                'State': {'Running': False}, 'Config': {'Env': ['SECRET=private']},
                'HostConfig': {'PortBindings': {'8002/tcp': [{'HostIp':'127.0.0.1','HostPort':'8002'}]}}}
                for value in args[5:]])
        with patch.object(resources, 'command', side_effect=command):
            result = resources.native_port_inventory()
        self.assertTrue(result['truncated'])
        self.assertEqual(len(result['containers']), 128)
        self.assertNotIn('SECRET', str(result))
        self.assertTrue(all(c[1] in ('ps', 'inspect') for c in calls))

    def test_missing_docker_or_gpu_is_visible(self):
        with patch.object(resources.platform, 'system', return_value='Linux'), patch.object(resources, 'command', side_effect=OSError('unavailable')):
            result = resources.collect({})
        self.assertIn('NVIDIA hardware unavailable', result['errors'])
        self.assertIn('Docker storage inspection unavailable', result['errors'])

    def test_recipe_checks_bind_candidates_and_keep_missing_proof_separate(self):
        import json
        ids = [format(i, '064x') for i in range(6)]
        image = 'sha256:' + 'b' * 64
        def command(*args):
            if args[1] == 'ps': return '\n'.join(ids)
            return json.dumps([{'Id': cid, 'Image': image, 'State': {'Running': False},
                'HostConfig': {'PortBindings': {'8002/tcp': [{'HostPort': '8002'}]}}} for cid in ids])
        calls = []
        def probe(cid, selected_image):
            calls.append((cid, selected_image))
            if cid == ids[0]:
                return {'state': 'verified', 'container': cid, 'image': selected_image,
                        'container_state_unchanged': True, 'supported': {'sampler_mode': ['heun']}}
            if cid == ids[1]: raise OSError('private Docker error and secret path')
            if cid == ids[2]: return {'state': 'verified', 'container': ids[0], 'image': image, 'container_state_unchanged': True}
            return {'state': 'verified', 'container': cid, 'image': image, 'container_state_unchanged': False}
        with patch.object(resources, 'command', side_effect=command):
            result = resources.native_port_inventory(probe)
        self.assertEqual(calls, [(cid, image) for cid in ids[:4]])
        self.assertEqual([r['recipe_contract']['state'] for r in result['containers']],
                         ['verified', 'unverified', 'unverified', 'unverified', 'not_checked', 'not_checked'])
        self.assertEqual(result['state'], 'observed')
        self.assertNotIn('secret', str(result))
        self.assertTrue(all(r['running'] is False for r in result['containers']))

    def test_video_port_does_not_trigger_ace_source_reads(self):
        import json
        cid = 'a' * 64
        def command(*args):
            if args[1] == 'ps': return cid
            return json.dumps([{'Id': cid, 'Image': 'sha256:' + 'b'*64,
                'State': {'Running': False}, 'HostConfig': {'PortBindings': {'8188/tcp': [{'HostPort': '8188'}]}}}])
        with patch.object(resources, 'command', side_effect=command):
            result = resources.native_port_inventory(lambda *_: self.fail('Video is not an ACE candidate'))
        self.assertNotIn('recipe_contract', result['containers'][0])


if __name__ == '__main__':
    unittest.main()

File: ds4-gateway/media-resources.test.mjs
import test from 'node:test';
import assert from 'node:assert/strict';
import {createMediaResources,mediaRecipeResources,inspectMediaResources} from './media-resources.mjs';
import {createMediaTools} from './genie-media.mjs';
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';

test('media resources keep permission, observations and fit claims separate',async()=>{
  let enabled=true,calls=0,fail=false,release;
  const waiting=new Promise(r=>release=r),target={container:'a'.repeat(64),ssh:['fixture']};
  const resources=createMediaResources({genie_chat:{inspection:{workers:{one:target}}}},{isEnabled:()=>enabled,inspect:async t=>{calls++;assert.equal(t,target);await waiting;if(fail)throw Error('SSH unavailable');return {observed_at:'2026-09-16T00:00:00Z',system:'Linux',architecture:'aarch64',docker_architecture:'aarch64',gpu_names:['NVIDIA GB10'],memory_total_bytes:128*2**30,memory_available_bytes:10*2**30,disks:[{location:'home',free_bytes:200*2**30}],errors:[]};}});
  const tools=createMediaTools({read:async()=>({jobs:[],hosts:[{id:'one',engines:[{id:'h3',enrolled:true,allowed:true}]}]}),resources});
  const first=tools.tool({action:'inspect',worker_id:'one'}),second=tools.tool({action:'inspect',worker_id:'one'});release();
  const [a,b]=await Promise.all([first,second]);assert.deepEqual(a,b);assert.equal(calls,1);assert.equal(a.recipe_platform_matches,true);assert.match(a.setup,/native qualification/);assert.equal(a.recipes.find(r=>r.engine==='h3').model_bytes_required,63440965087);assert.equal(mediaRecipeResources.find(r=>r.engine==='ace-step').model_bytes_required,28496774102);
  assert.equal((await tools.tool({action:'status'})).resource_checks.one.state,'observed');
  assert.equal(a.recipes[0].installed_model_inventory_checked,false);assert.equal(a.existing_engines[0].enrolled,true);assert.match(a.lifecycle,/do not need simultaneous residency/);
  await assert.rejects(tools.tool({action:'inspect',worker_id:'not-enrolled'}),/registered/);
  await assert.rejects(tools.tool({action:'inspect',worker_id:'one',command:'stop'}));
  enabled=false;await assert.rejects(tools.tool({action:'inspect',worker_id:'one'}),/switched off/);assert.equal(calls,1);
  enabled=true;fail=true;await assert.rejects(tools.tool({action:'inspect',worker_id:'one'}),/SSH unavailable/);assert.equal(resources.status().one.state,'unavailable');
});

test('other platforms remain unqualified without changing existing capabilities',async()=>{
  const resources=createMediaResources({genie_chat:{inspection:{workers:{mac:{kind:'omlx-local'}}}}},{inspect:async()=>({system:'Darwin',architecture:'arm64',gpu_names:[],docker_architecture:null})});
  const result=await resources.inspect('mac');assert.equal(result.recipe_platform_matches,false);assert.match(result.setup,/Existing engine enrollments and serving capabilities are unchanged/);
});

test('resource observations distinguish the two physical members of an enrolled pair',async()=>{
 const worker={id:'pair',url:'http://fixture'},config={workers:[worker],genie_chat:{inspection:{workers:{pair:{ssh:['head'],container:'head-model'}}}},media_jobs:{pairs:{pair:{kind:'glm53-docker-pair',model:'GLM',worker_binding:{...worker},members:[{ssh:'head',container:'head-model'},{ssh:'rank',container:'rank-model'}]}}}};
 const seen=[],service=createMediaResources(config,{inspect:async target=>{seen.push(target);return {system:'Linux',architecture:'aarch64',docker_architecture:'arm64',gpu_names:['GB10']};}});
 await service.inspect('pair',0);await service.inspect('pair',1);
 assert.deepEqual(seen.map(t=>t.ssh[0]),['head','rank']);assert.equal(service.status()['pair:1'].member,1);
 await assert.rejects(service.inspect('pair',2),/enrolled paired/);
 config.media_jobs.pairs.pair.worker_binding.url='http://changed';await assert.rejects(service.inspect('pair',1),/enrolled paired/);
});

test('packaged SSH collector reads the exact stopped ACE witness without running its CLI or native commands',async t=>{
 const folder=fs.mkdtempSync(path.join(os.tmpdir(),'media-resource-transport-')),oldPath=process.env.PATH;
 t.after(()=>{process.env.PATH=oldPath;fs.rmSync(folder,{recursive:true,force:true});});
 const script=`#!/usr/bin/env python3
import json,platform,shlex,subprocess,sys
from pathlib import Path
cid='a'*64
image='sha256:'+'b'*64
calls=[]
platform.system=lambda:'Linux'
platform.machine=lambda:'aarch64'
def output(args,**kwargs):
 calls.append(args)
 if args[0]=='nvidia-smi':return 'NVIDIA GB10'
 if args[:2]==('docker','info'):return json.dumps({'Architecture':'aarch64'})
 if args[:2]==('docker','ps'):return cid
 if args[:2]==('docker','inspect'):return json.dumps([{'Id':cid,'Image':image,'State':{'Running':False},'HostConfig':{'PortBindings':{'8002/tcp':[{'HostPort':'8002'}]}},'Mounts':[]}])
 raise AssertionError('Unexpected resource command')
def run(args,**kwargs):
 calls.append(args)
 if args==['docker','inspect',cid]:return subprocess.CompletedProcess(args,0,json.dumps([{'Id':cid,'Image':image,'Config':{'WorkingDir':'/opt/ace-step'},'State':{'Running':False}}]).encode())
 if args==['docker','cp',cid+':/opt/stargate/recipe-fields-verification.json','-']:raise subprocess.CalledProcessError(1,args,stderr=b'private path must not leak')
 raise AssertionError('Unexpected native command')
subprocess.check_output=output
subprocess.run=run
args=shlex.split(sys.argv[-1])
assert args[:4]==['python3','-I','-B','-c']
exec(compile(args[4],'<fixed-collector>','exec'),{'__name__':'__main__'})
Path(__file__).with_name('calls.json').write_text(json.dumps(calls))
`;
 fs.writeFileSync(path.join(folder,'ssh'),script,{mode:0o700});process.env.PATH=folder+path.delimiter+oldPath;
 const result=await inspectMediaResources({container:'llm',ssh:['fixture']});
 assert.equal(result.native_port_inventory.state,'observed');
 const [candidate]=result.native_port_inventory.containers;
 assert.equal(candidate.running,false);assert.equal(candidate.recipe_contract.state,'unverified');
 assert.ok(!JSON.stringify(result).includes('private path'));
 const calls=JSON.parse(fs.readFileSync(path.join(folder,'calls.json')));
 assert.ok(calls.some(c=>c[1]==='cp'&&c[2]==='a'.repeat(64)+':/opt/stargate/recipe-fields-verification.json'));
 assert.ok(calls.every(c=>c[0]==='nvidia-smi'||c[0]==='docker'&&['ps','info','inspect','cp'].includes(c[1])));
});
